### ISSUE TITLE
"anonymous types declared in an anonymous union are an extension [-Wnested-anon-types]"

### ISSUE THREAD

**Issue by anntzer** [pybind/pybind11#1204]
## Issue description

Compiling a file that includes pybind11/detail/common.h (it is private but obviously included by other public headers) with clang -Wpedantic triggers the warning
```
In file included from src/main.cpp:1:
In file included from /usr/include/python3.6m/pybind11/pybind11.h:43:
In file included from /usr/include/python3.6m/pybind11/attr.h:13:
In file included from /usr/include/python3.6m/pybind11/cast.h:13:
In file included from /usr/include/python3.6m/pybind11/pytypes.h:12:
/usr/include/python3.6m/pybind11/detail/common.h:386:9: warning: anonymous types declared in an anonymous union are an extension [-Wnested-anon-types]
        struct 
```

## Reproducible example code

Add `#include <pybind11/detail/common.h>` to python_example and compile with
```
CC=clang CFLAGS=-Wpedantic CXX=clang setup.py build_ext
```

**Comment by Psirus** [comment]
I've run into this as well. Since I'm using the provided CMake macros (thank you for this, btw), I would be fine with a solution like this:

~~~diff
diff --git a/tools/pybind11Tools.cmake b/tools/pybind11Tools.cmake
index a7c471a..b8bfb1c 100644
--- a/tools/pybind11Tools.cmake
+++ b/tools/pybind11Tools.cmake
@@ -130,7 +130,7 @@ function(pybind11_add_module target_name)
 
   add_library(${target_name} ${lib_type} ${exclude_from_all} ${ARG_UNPARSED_ARGUMENTS})
 
-  target_include_directories(${target_name}
+  target_include_directories(${target_name} SYSTEM
     PRIVATE ${PYBIND11_INCLUDE_DIR}  # from project CMakeLists.txt
     PRIVATE ${pybind11_INCLUDE_DIR}  # from pybind11Config
     PRIVATE ${PYTHON_INCLUDE_DIRS})
~~~

**Comment by Psirus** [comment]
Was just a suggestion in case getting the headers warning free was too intrusive. Obviously I'd prefer a solution that works for everyone as well.

**Comment by wjakob** [comment]
Thank you -- yes, please do.

**Comment by jagerman** [comment]
Fixed by #1248
